cdap-ranger/cdap-ranger-binding/src/main/java/co/cask/cdap/security/authorization/ranger/binding/RangerAuthorizer.java

```diff
@@ -0,0 +1,301 @@
+  private static final String KEY_ARTIFACT = "artifact";
+  private static final String KEY_APPLICATION = "application";
+  private static final String KEY_DATASET = "dataset";
+  private static final String KEY_STREAM = "stream";
+  private static final String KEY_PROGRAM = "program";
+  private static final String KEY_DATASET_MODULE = "dataset_module";
+  private static final String KEY_DATASET_TYPE = "dataset_type";
+  private static final String KEY_SECUREKEY = "securekey";
+
+  // using # as we don't allow it in entity names
+  private static final String RESOURCE_SEPARATOR = "#";
+
+  private static volatile RangerBasePlugin rangerPlugin = null;
+  private AuthorizationContext context;
+  // cdap instance name
+  private String instanceName;
+
+  @Override
+  public synchronized void initialize(AuthorizationContext context) throws Exception {
+    this.context = context;
+    Properties properties = context.getExtensionProperties();
+    instanceName = properties.containsKey(Constants.INSTANCE_NAME) ?
+      properties.getProperty(Constants.INSTANCE_NAME) : "cdap";
+    if (rangerPlugin == null) {
+      try {
+        UserGroupInformation ugi = UserGroupInformation.getLoginUser();
+        Preconditions.checkNotNull(ugi, "Kerberos login information is not available. UserGroupInformation is null");
+        // set the login user as the user as whom cdap is running as this is needed for kerberos authentication
+        MiscUtil.setUGILoginUser(ugi, null);
+        LOG.debug("Initializing Ranger CDAP Plugin with UGI {}", ugi);
+      } catch (Throwable t) {
+        LOG.error("Error getting principal.", t);
+      }
+      // the string name here should not be changed as this uniquely identifies the plugin in ranger. If it's
+      // changed it will require changing all the supporting xml file which is in this package.
+      rangerPlugin = new RangerBasePlugin("cdap", "cdap");
+    }
+    rangerPlugin.init();
+    RangerDefaultAuditHandler auditHandler = new RangerDefaultAuditHandler();
+    rangerPlugin.setResultProcessor(auditHandler);
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    LOG.debug("Enforce called on entity {}, principal {}, action {}", entity, principal, action);
+    if (rangerPlugin == null) {
+      throw new RuntimeException("CDAP Ranger Authorizer is not initialized.");
+    }
+
+    if (principal.getType() != Principal.PrincipalType.USER) {
+      throw new IllegalArgumentException(String.format("The principal type for current enforcement request is '%s'. " +
+                                                         "Authorization enforcement is only supported for '%s'.",
+                                                       principal.getType(), Principal.PrincipalType.USER));
+    }
+    String requestingUser = principal.getName();
+    String ip = InetAddress.getLocalHost().getHostName();
```

Comment:
AuthenticationContext does not have it as of now. If we need to use that we will need to do some changes in CDAP to propagate the ip here.